docs/design-docs/07-package-orchestration.md

```diff
@@ -0,0 +1,491 @@
+(k8s apiserver and Config Sync).
+
+![](./Porch%20Architecture.svg)
+
+In addition to satisfying requirements highlighted above, the focus of the
+architecture was to:
+
+* establish clear components and interfaces
+* support a low-latency package authoring experience required by the UIs
+
+The Porch components are:
+
+#### Porch Server
+
+The Porch server is implemented as [Kubernetes extension API server][apiserver].
+The benefits of using Kubernetes extension API server are:
+
+* well-defined and familiar API style
+* availability of generated clients
+* integration with existing Kubernetes ecosystem and tools such as `kubectl`
+  CLI, [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)
+* avoids requirement to open another network port to access a separate endpoint
+  runninig inside k8s cluster (this is a distinct advantage over gRPC which we
+  considered as an alternative approach)
+
+Resources implemented by Porch include:
+
+* `PackageRevision` - represents the _metadata_ of the configuration package
+  revision stored in a _package_ repository.
+* `PackageRevisionResources` - represents the _contents_ of the package revision
+* `Function` - represents a [KRM function][krm functions] discovered in
+  a registered _function_ repository.
+
+Note that each configuration package revision is represented by a _pair_ of
+resources which each present a different view (or [representation][] of the same
+underlying package revision.
+
+Repository registration is supported by a `Repository` [custom resource][crds].
+
+**Porch server** itself comprises several key components, including:
+
+* The *Porch aggregated apiserver* which implements the integration into the
+  main Kubernetes apiserver, and directly serves API requests for the
+  `PackageRevision`, `PackageRevisionResources` and `Function` resources.
+* Package orchestration *engine* which implements the package lifecycle
+  operations, and package mutation workflows
+* *CaD Library* which implements specific package manipulation algorithms such
+  as package rendering (evaluation of package's function *pipeline*),
+  initialization of a new package, etc. The CaD Library is shared with `kpt`
+  where it likewise provides the core package manipulation algorithms.
+* *Package cache* which enables both local caching, as well as abstract
+  mainpulation of packages and their contents irrespectively of the underlying
+  storage mechanism (Git, or OCI)
+* *Repository adapters* for Git and OCI which implement the specific logic of
+  interacting with those types of package repositories.
+* *Function runtime* which implements support for evaluating
+  [kpt functions][functions] and multi-tier cache of functions to support
+  low latency function evaluation
+
+#### Function Runner
+
+**Function runner** is a separate service responsible for evaluating
+[kpt functions][functions]. Function runner exposes a [gRPC](https://grpc.io/)
```

Comment:
Maybe explain here why the function runner exposes a gRPC interface instead of a KRM interface. Ref the section about why KRM was chosen for the porch server.

Reply:
thanks! added rationale. ptal.